core/src/service/sync_service.rs

```diff
@@ -24,122 +28,251 @@ pub struct WorkCalculated {
+            last_synced: self.tx.last_synced.get(&OneKey {}).map(|s| *s as u64),
+            base: (&mut self.tx.base_metadata).to_staged(Vec::new()),
+            local: (&mut self.tx.local_metadata).to_staged(Vec::new()),
+            username_by_public_key: &mut self.tx.username_by_public_key,
+            public_key_by_username: &mut self.tx.public_key_by_username,
+            client: self.client,
+            config: self.config,
+        };
 
-        Ok(())
+        let update_as_of = sync_context.sync(&mut update_sync_progress)?;
+
+        if let Some(root) = sync_context.root {
+            self.tx.root.insert(OneKey {}, root);
+        }
+        if let Some(last_synced) = sync_context.last_synced {
+            self.tx.last_synced.insert(OneKey {}, last_synced as i64);
+        }
+        sync_context.base.to_lazy().promote();
+        sync_context.local.to_lazy().promote();
+
+        Ok(WorkCalculated { work_units, most_recent_update_from_server: update_as_of as u64 })
+    }
+}
+
+struct SyncContext<'a, 'b, Base, Local, UsernameByPublicKey, PublicKeyByUsername, Client>
+where
+    Base: TreeLike<F = SignedFile>,
+    Local: TreeLike<F = SignedFile>,
+    UsernameByPublicKey: SchemaEvent<Owner, String>,
+    PublicKeyByUsername: SchemaEvent<String, Owner>,
+    Client: Requester,
+{
+    // when dry_run is true, sync skips document downloads/uploads and disk reads/writes
+    dry_run: bool,
+
+    // root, last_synced, base metadata, and local metadata have values pre-loaded, are read from
+    // and written to here, and are "committed" if we are not doing a dry run
+    root: Option<Uuid>,
+    last_synced: Option<u64>,
+    base: StagedTree<Base, Vec<SignedFile>>,
+    local: StagedTree<Local, Vec<SignedFile>>,
+
+    // public key cache is written to, dry run or not
+    username_by_public_key: &'a mut TransactionTable<'b, Owner, String, UsernameByPublicKey>,
+    public_key_by_username: &'a mut TransactionTable<'b, String, Owner, PublicKeyByUsername>,
+
+    // account pre-loaded for convenience
+    account: Account,
+
+    // client for network requests and config for disk i/o
+    client: &'a Client,
+    config: &'a Config,
+}
+
+impl<'a, 'b, Base, Local, UsernameByPublicKey, PublicKeyByUsername, Client>
+    SyncContext<'a, 'b, Base, Local, UsernameByPublicKey, PublicKeyByUsername, Client>
+where
+    Base: TreeLike<F = SignedFile>,
+    Local: TreeLike<F = SignedFile>,
+    UsernameByPublicKey: SchemaEvent<Owner, String>,
+    PublicKeyByUsername: SchemaEvent<String, Owner>,
+    Client: Requester,
+{
+    fn sync<F>(&mut self, report_sync_operation: &mut F) -> CoreResult<i64>
+    where
+        F: FnMut(SyncOperation),
+    {
+        self.prune()?;
+        let update_as_of = self.pull(report_sync_operation)?;
+        self.last_synced = Some(update_as_of as u64);
+        self.push_metadata(report_sync_operation)?;
+        self.push_documents(report_sync_operation)?;
+        Ok(update_as_of)
```

Comment:
a couple of extraneous calls to prune and pull were removed now that server does not apply implicit file deletions or anything like that